### PR TITLE
Enable compilation tests for microbenchmarks

### PR DESCRIPTION
**Pull request by hanhanW**

### CODE REVIEW

**Review by MaheshRavishankar — APPROVED** [review]

**Comment by hanhanW** [comment]
@pzread you can consider to add microbenchmarks to this folder. For convolution cases, we don't have to start with mhlo level. We can add linalg_conv_xxx.mlir files instead.

**Review by antiagainst — CHANGES_REQUESTED** [review]
Wait. I think this will by default put all microbencmarks to tests and gotten invoked when we `ctest`? Can we make these opt-in and not run by default when running tests?

**Comment by hanhanW** [comment]
> Wait. I think this will by default put all microbencmarks to tests and gotten invoked when we `ctest`? Can we make these opt-in and not run by default when running tests?

Yes, you're right. It's sounds good to me. How do I make them opt-in and not run by default?

**Comment by antiagainst** [comment]
Not sure ctest supports optionality.. Could use a CMake option like `IREE_BUILD_MICROBENCHMARKS` to guard against this I think.

**Comment by MaheshRavishankar** [comment]
> Wait. I think this will by default put all microbencmarks to tests and gotten invoked when we `ctest`? Can we make these opt-in and not run by default when running tests?

Just curious why not run these tests? As far as I understand, they are just making sure these compile and run, and it should be a small fraction of the  time taken by other tests (like the model tests). So I think it is OK to be added by default?

**Comment by hanhanW** [comment]
After thinking a while, maybe it's fine to put them to default ctest? The purpose is to test compilation and runtime for the ops. We're not running microbenchmarks in ctest.

**Comment by antiagainst** [comment]
Time to run these microbenchmarks is indeed one concern. For stable benchmarks we tend to include large-shaped ops. Also we want to consider many variants. It can adds up quickly. Looks [right now](https://source.cloud.google.com/results/invocations/9f8f6d63-80b4-4192-89f6-188e37fc6d7c/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log) the existing ones is taking ~100s, while model tests taking ~150s. So they are at similar levels of latency already. I think we are looking to include further stuff here.

More importantly (micro)benchmarks are not tests. Tests are meant for correctness (including those model tests). Each checks a different feature or from a different angle. Typically we don't check in multiple tests covering the same thing. Microbenchmarks actually does that by just focusing one single op and varying its shape. There is little additional coverage here to warrant always running them in each `ctest` invocation for everybody during daily development.

I do see the needs of keeping these IRs up to date and able to compile though. (Given that there is no verifying the output at all right now, I'm not sure we need to actually run them.) It happens at a quite low frequency and only after some integration. If we want to guard against such stale IR cases, using CI is enough for it IMO. So making it optional and default to off and enable for CIs would cover that, and we don't need to pay the cost during daily development after each `ctest` invocation. :)

**Comment by MaheshRavishankar** [comment]
> Time to run these microbenchmarks is indeed one concern. For stable benchmarks we tend to include large-shaped ops. Also we want to consider many variants. It can adds up quickly. Looks [right now](https://source.cloud.google.com/results/invocations/9f8f6d63-80b4-4192-89f6-188e37fc6d7c/targets/iree%2Fgcp_ubuntu%2Fcmake%2Flinux%2Fx86-swiftshader%2Fpresubmit/log) the existing ones is taking ~100s, while model tests taking ~150s. So they are at similar levels of latency already. I think we are looking to include further stuff here.
> 
> More importantly (micro)benchmarks are not tests. Tests are meant for correctness (including those model tests). Each checks a different feature or from a different angle. Typically we don't check in multiple tests covering the same thing. Microbenchmarks actually does that by just focusing one single op and varying its shape. There is little additional coverage here to warrant always running them in each `ctest` invocation for everybody during daily development.
> 
> I do see the needs of keeping these IRs up to date and able to compile though. (Given that there is no verifying the output at all right now, I'm not sure we need to actually run them.) It happens at a quite low frequency and only after some integration. If we want to guard against such stale IR cases, using CI is enough for it IMO. So making it optional and default to off and enable for CIs would cover that, and we don't need to pay the cost during daily development after each `ctest` invocation. :)

These are indeed meant for correctness (and by the above argument should run CIs). The point is these micro benchmarks should always compile + run. Right now these are not "testing the result" cause they are bigger size inputs and outputs and testing the result for these sizes is hard. I think it is useful to have the test compile + run (run is just to make sure it doesnt say segfault at runtime or something like that) even if these tests dont have correctness included. 

100s is a lot actually, but it is not that much. I think its worth having these.

**Comment by dcaballe** [comment]
Interestingly @MaheshRavishankar, @hanhanW and I had exactly the same discussion a few days ago :).  I think I align with what @antiagainst said: testing and benchmarking should be "independent". Otherwise, testing will impose unnecessary runtime constraints on what we can benchmark. Running the benchmarks won't provide too much value from a testing perspective since we are not validating the output. Yes, we would know if the benchmark is crashing or not but that's only marginal information. We can mess up the IR in many ways that won't lead to a runtime crash. What I've seen in other projects is that benchmarks are accompanied by representative e2e tests where the output is checked and that run within a more reasonable time budget. Even e2e tests may become problematic: https://discourse.llvm.org/t/check-mlir-times-examples-standalone-testing-time/6073 :)

I think enabling benchmarks with a cmake flag could be a good first step, as suggested above.

Hopefully that helps!

**Comment by MaheshRavishankar** [comment]
> Interestingly @MaheshRavishankar, @hanhanW and I had exactly the same discussion a few days ago :). I think I align with what @antiagainst said: testing and benchmarking should be "independent". Otherwise, testing will impose unnecessary runtime constraints on what we can benchmark. Running the benchmarks won't provide too much value from a testing perspective since we are not validating the output. Yes, we would know if the benchmark is crashing or not but that's only marginal information. We can mess up the IR in many ways that won't lead to a runtime crash. What I've seen in other projects is that benchmarks are accompanied by representative e2e tests where the output is checked and that run within a more reasonable time budget. Even e2e tests may become problematic: [discourse.llvm.org/t/check-mlir-times-examples-standalone-testing-time/6073](https://discourse.llvm.org/t/check-mlir-times-examples-standalone-testing-time/6073) :)
> 
> I think enabling benchmarks with a cmake flag could be a good first step, as suggested above.
> 
> Hopefully that helps!

Well there is performance tracking and there is testing. They are indeed independent, but making sure things that are needed for performance tracking are "not broken" should very much be part of testing. We could maybe add this as part of post-submit testing if 100s of testing is too onerous (no one needs/should run the test all the time. You probably need to run it once locally to verify, or just let CI run it which is what I do, so 100s is very doable). Right now we are relying on post-submit to catch places where things are broken and either revert or fix forward on those errors. We could do that for these tests too if there is a place in the post-submit CI these could slot in.
Re: full correctness testing, of course that would be better, but is hard to do for large tensors. At least a runtime crash tells you something is broken (like maybe too much stack allocation, etc.). Hopefully other smaller unit tests are enough to catch correctness issues. 

I still dont see a reason to not have these tests checked in. Maybe post-submit is an option.

**Comment by hanhanW** [comment]
I understand that putting this to default adds overheads for everybody during daily development. It makes sense to have an option to control it. People who care it can turn on the option themselves. We can ask CI to test it.

The current revision only enables compilation, but not runtime. It's okay to me because I believe they are all covered in our unit tests and integration tests. We can add runtime tests to CI if we run into issues.

Do we have preference to run it on which CI bot?
I could not find how to put it to other presubmit bots, so I temporary add it to `linux/x86-swiftshader-asan`.

**Review by antiagainst — APPROVED** [review]
LGTM, just some additional nits.

File: build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-swiftshader-asan/build.sh [hunk 1]
```diff
@@ -37,6 +37,9 @@ CMAKE_ARGS=(
   "-DIREE_ENABLE_ASAN=ON"
   "-B" "${CMAKE_BUILD_DIR?}"
 
+  # Also check if microbenchmarks are buildable.
+  "-DIREE_BUILD_MICROBENCHMARKS=ON"
```
[the remark this answers is not in the document]
Comment (antiagainst) [reply build_tools/kokoro/gcp_ubuntu/cmake/linux/x86-swiftshader-asan/build.sh:41]: `cmake/build_android_benchmark.sh` runs on pull requests.

**Comment by antiagainst** [comment]
> I still dont see a reason to not have these tests checked in. Maybe post-submit is an option.

Having them checked by CI is fine to me; either presubmit or postsubmit works. I was saying it's better to allow the optionality so that folks can choose to avoid running these in the daily development workflow, as they are not the same as the other tests and we have correctness coverage for these ops in e2e/ directory. :)

**Comment by MaheshRavishankar** [comment]
> > I still dont see a reason to not have these tests checked in. Maybe post-submit is an option.
> 
> Having them checked by CI is fine to me; either presubmit or postsubmit works. I was saying it's better to allow the optionality so that folks can choose to avoid running these in the daily development workflow, as they are not the same as the other tests and we have correctness coverage for these ops in e2e/ directory. :)

Understood. I guess personally I dont run into the issue since I have vscode setup so I can run different test suites. So I rarely run all tests (mostly cause we have a resnet test in the e2e/models that takes as much time as all other tests combined in debug builds)

**Pull request merged**